src/NHibernate/Action/BulkOperationCleanupAction.cs

```diff
@@ -14,34 +14,49 @@ namespace NHibernate.Action
 	/// Implementation of BulkOperationCleanupAction.
 	/// </summary>
 	[Serializable]
-	public partial class BulkOperationCleanupAction : IAsyncExecutable, IAfterTransactionCompletionProcess
+	public partial class BulkOperationCleanupAction :
+		IAsyncExecutable,
+		IAfterTransactionCompletionProcess,
+		ICacheableExecutable
 	{
-		private readonly ISessionImplementor session;
-		private readonly HashSet<string> affectedEntityNames = new HashSet<string>();
-		private readonly HashSet<string> affectedCollectionRoles = new HashSet<string>();
-		private readonly List<string> spaces;
+		private readonly ISessionFactoryImplementor _factory;
+		private readonly HashSet<string> _affectedEntityNames;
+		private readonly HashSet<string> _affectedCollectionRoles;
+		private readonly string[] _spaces;
+		private readonly bool _hasCache;
 
 		public BulkOperationCleanupAction(ISessionImplementor session, IQueryable[] affectedQueryables)
 		{
-			this.session = session;
-			List<string> tmpSpaces = new List<string>();
-			for (int i = 0; i < affectedQueryables.Length; i++)
+			_factory = session.Factory;
+			var tmpSpaces = new HashSet<string>();
+			foreach (var queryables in affectedQueryables)
 			{
-				if (affectedQueryables[i].HasCache)
+				if (queryables.HasCache)
 				{
-					affectedEntityNames.Add(affectedQueryables[i].EntityName);
+					_hasCache = true;
+					if (_affectedEntityNames == null)
+					{
+						_affectedEntityNames = new HashSet<string>();
+					}
+
+					_affectedEntityNames.Add(queryables.EntityName);
 				}
-				ISet<string> roles = session.Factory.GetCollectionRolesByEntityParticipant(affectedQueryables[i].EntityName);
+
+				var roles = _factory.GetCollectionRolesByEntityParticipant(queryables.EntityName);
```

Comment:
Collection persister can also have cache. But currently `_hasCache` is only calculated for entity persister. And shouldn't we retrieve here list for only cached collection roles?